test/loader_tests.js

```diff
@@ -0,0 +1,333 @@
+          return deps.dep;
+        }
+      }
+    })();
+
+    assume(called).is.false();
+    assume(await load('base')).equals(a);
+    assume(called).is.true();
+  });
+
+  it('should load a simple virtual dependency', async function() {
+    let a = {a: 1};
+    let called = false;
+    
+    let baseLoader = subject({
+      dep: {
+        setup: () => {
+          return 2;
+        }
+      },
+      base: {
+        requires: ['dep', 'dep2'],
+        setup: async deps => {
+          assume(a).equal(await deps.dep2);
+          return deps.dep2;
+        }
+      }
+    }, ['dep2']);
+
+    assume(called).is.false();
+    let load = baseLoader({
+      dep2: {
+        setup: () => {
+          called = true;
+          return a;
+        },
+      },
+    });
+    assume(await load('base')).equals(a);
+    assume(called).is.true();
+  });
+
+  it('should detect and bail on cyclic dependency', async function() {
+    try {
+      let load = subject({
+        dep1: {
+          requires: ['dep2'],
+          setup: () => true
+        },
+        dep2: {
+          requires: ['dep3'],
+          setup: () => true
+        },
+        dep3: {
+          requires: ['dep1'],
+          setup: () => true
+        },
+        base: {
+          requires: ['dep1'],
+          setup: () => true
+        },
+      })({});
+      await load('base');
+      throw new Error('this should not work');
+    } catch (e) {
+      if (!e.message.match(/^Cyclical dependency:/)) {
+        throw e;
+      }
+    }
+  });
+
+  it('should load different types of static dependencies', async function() {
+    let a = {a: 1};
+    let b = {b: 2};
+    let c = {c: 2};
+    let baseLoader = subject({
+      string: 'a-string',
+      object: a,
+      number: 123.456,
+      promise: new Promise(res => { res(b) }),
+      func: () => { return c },
+    }, ['base']);
+
+    let load = baseLoader({
+      base: {
+        requires: ['string', 'object', 'number', 'promise', 'func'],
+        setup: async deps => {
+          assume(await deps.string).equals('a-string');
+          assume(await deps.object).equals(a);
+          assume(await deps.number).equals(123.456);
+          assume(await deps.promise).equals(b);
+          assume((await deps.func).call()).equals(c);
+        },
+      },
+    });
+
+    await load('base');
+  });
+
+  it('should work with a complex dependency graph', async function() {
+    let load = subject({
+      dep1: {
+        requires: ['dep2', 'dep3'],
+        setup: () => true
+      },
+      dep2: {
+        requires: ['dep4'],
+        setup: () => true
+      },
+      dep3: {
+        requires: ['dep4'],
+        setup: () => true
+      },
+      dep4: {
+        requires: [],
+        setup: () => true
+      },
+      staticDep1: 'john',
+      base: {
+        requires: ['dep1', 'staticDep1'],
+        setup: async deps => {
+          assume(await deps.staticDep1).equals('john');
```

Comment:
Okay... Maybe I'm wrong... But I think we can avoid the `await` here... I'll take a shot at it and see if I can't refactor it and make a test that shows it works... If not we can do this await...